apps/indexer/src/api/services/historical-balances/historical-balances.service.ts

```diff
@@ -0,0 +1,272 @@
+      throw new Error(`Chain with ID ${chainId} not found`);
+    }
+
+    this.client = createPublicClient({
+      chain,
+      transport: http(rpcUrl),
+    });
+  }
+
+  /**
+   * Fetches historical balances for multiple addresses at a specific block
+   * Uses Multicall3 for efficient batch queries when available (block 19+)
+   * Falls back to individual calls for earlier blocks
+   */
+  async getHistoricalBalances({
+    addresses,
+    blockNumber,
+    daoId,
+  }: HistoricalBalancesRequest): Promise<HistoricalBalance[]> {
+    const tokenAddress = this.getTokenAddress(daoId);
+
+    if (!tokenAddress) {
+      throw new Error(`Token address not found for DAO: ${daoId}`);
+    }
+
+    try {
+      // Use Multicall3 for blocks 19+ (after Multicall3 deployment), individual calls for earlier blocks
+      if (blockNumber >= 19) {
+        return await this.getBalancesWithMulticall(addresses, blockNumber, tokenAddress);
+      } else {
+        return await this.getBalancesIndividually(addresses, blockNumber, tokenAddress);
+      }
+    } catch (error) {
+      console.error("Error fetching historical balances:", error);
+      // Fallback to individual calls if multicall fails
+      console.log("Falling back to individual calls...");
+      return await this.getBalancesIndividually(addresses, blockNumber, tokenAddress);
+    }
+  }
+
+  /**
+   * Get balances using Multicall3 for efficient batch processing
+   */
+  private async getBalancesWithMulticall(
+    addresses: Address[],
+    blockNumber: number,
+    tokenAddress: Address
+  ): Promise<HistoricalBalance[]> {
+    // Prepare multicall data
+    const calls = addresses.map((address) => {
+      const calldata = encodeFunctionData({
+        abi: [
+          {
+            inputs: [{ name: "account", type: "address" }],
+            name: "balanceOf",
+            outputs: [{ name: "", type: "uint256" }],
+            stateMutability: "view",
+            type: "function",
+          },
+        ] as const,
+        functionName: "balanceOf",
+        args: [address],
+      });
+
+      return {
+        target: tokenAddress,
+        callData: calldata,
+      };
+    });
+
+    // Call Multicall3.aggregate
+    const result = await readContract(this.client, {
+      address: MULTICALL3_ADDRESS,
+      abi: [
+        {
+          inputs: [
+            {
+              components: [
+                { name: "target", type: "address" },
+                { name: "callData", type: "bytes" },
+              ],
+              name: "calls",
+              type: "tuple[]",
+            },
+          ],
+          name: "aggregate",
+          outputs: [
+            { name: "blockNumber", type: "uint256" },
+            { name: "returnData", type: "bytes[]" },
+          ],
+          stateMutability: "payable",
+          type: "function",
+        },
+      ] as const,
+      functionName: "aggregate",
+      args: [calls],
+      blockNumber: BigInt(blockNumber),
+    });
```

Comment:
why no to use the viem's [multicall](https://viem.sh/docs/contract/multicall.html) method? it has the [address of the contract on each chain](https://github.com/wevm/viem/blob/main/src/chains/definitions/arbitrum.ts#L20) (except anvil)